pkg/solver/matcher/match.go

```diff
@@ -0,0 +1,256 @@
+package matcher
+
+import (
+	"strings"
+
+	"github.com/lilypad-tech/lilypad/pkg/data"
+	"github.com/rs/zerolog/log"
+)
+
+type matchResult interface {
+	matched() bool
+	message() string
+}
+
+type offersMatched struct {
+	resourceOffer data.ResourceOffer
+	jobOffer      data.JobOffer
+}
+
+func (_ offersMatched) matched() bool   { return true }
+func (_ offersMatched) message() string { return "offers matched" }
+
+type cpuMismatch struct {
+	resourceOffer data.ResourceOffer
+	jobOffer      data.JobOffer
+}
+
+func (_ cpuMismatch) matched() bool   { return false }
+func (_ cpuMismatch) message() string { return "did not match CPU" }
+
+type gpuMismatch struct {
+	resourceOffer data.ResourceOffer
+	jobOffer      data.JobOffer
+}
+
+func (_ gpuMismatch) matched() bool   { return false }
+func (_ gpuMismatch) message() string { return "did not match GPU" }
+
+type ramMismatch struct {
+	resourceOffer data.ResourceOffer
+	jobOffer      data.JobOffer
+}
+
+func (_ ramMismatch) matched() bool   { return false }
+func (_ ramMismatch) message() string { return "did not match RAM" }
+
+type moduleIDError struct {
+	resourceOffer data.ResourceOffer
+	jobOffer      data.JobOffer
+	err           error
+}
+
+func (_ moduleIDError) matched() bool   { return false }
+func (_ moduleIDError) message() string { return "error getting module ID" }
+
+type moduleMismatch struct {
+	resourceOffer data.ResourceOffer
+	jobOffer      data.JobOffer
+}
+
+func (_ moduleMismatch) matched() bool   { return false }
+func (_ moduleMismatch) message() string { return "did not match modules" }
+
+type marketPriceUnavailable struct {
+	resourceOffer data.ResourceOffer
+	jobOffer      data.JobOffer
+}
+
+func (_ marketPriceUnavailable) matched() bool { return false }
+func (_ marketPriceUnavailable) message() string {
+	return "do not support market priced resource offers"
+}
+
+type priceMismatch struct {
+	resourceOffer data.ResourceOffer
+	jobOffer      data.JobOffer
+}
+
+func (_ priceMismatch) matched() bool { return false }
+func (_ priceMismatch) message() string {
+	return "fixed price job offer cannot afford resource offer"
+}
+
+type mediatorMismatch struct {
+	resourceOffer data.ResourceOffer
+	jobOffer      data.JobOffer
+}
+
+func (_ mediatorMismatch) matched() bool   { return false }
+func (_ mediatorMismatch) message() string { return "no matching mutual mediators" }
+
+type solverMismatch struct {
+	resourceOffer data.ResourceOffer
+	jobOffer      data.JobOffer
+}
+
+func (_ solverMismatch) matched() bool   { return false }
+func (_ solverMismatch) message() string { return "no matching solver" }
+
+// the most basic of matchers
+// basically just check if the resource offer >= job offer cpu, gpu & ram
+// if the job offer is zero then it will match any resource offer
+func matchOffers(
```

Comment:
Let's add some unit tests for this, i have a suspicion there might be a bug in this somewhere